envi/__init__.py

```diff
@@ -706,6 +709,108 @@ def getCallingConvention(self, name):
+        """
+        #FIXME: Handle endianness
+        bytes = self.readMemory(addr, size)
+        if bytes == None:
+            return None
+        if len(bytes) != size:
+            raise Exception("Read Gave Wrong Length At 0x%.8x (va: 0x%.8x wanted %d got %d)" % (self.getProgramCounter(),addr, size, len(bytes)))
+        if size == 1:
+            return struct.unpack("B", bytes)[0]
+        elif size == 2:
+            return struct.unpack(">H", bytes)[0]
+        elif size == 4:
+            return struct.unpack(">L", bytes)[0]
+        elif size == 8:
+            return struct.unpack(">Q", bytes)[0]
+
+    def writeMemValue(self, addr, value, size):
+        #FIXME change this (and all uses of it) to passing in format...
+        #FIXME: Remove byte check and possibly half-word check.  (possibly all but word?)
+        #FIXME: Handle endianness
+        if size == 1:
+            bytes = struct.pack("B",value & 0xff)
+        elif size == 2:
+            bytes = struct.pack(">H",value & 0xffff)
+        elif size == 4:
+            bytes = struct.pack(">L", value & 0xffffffff)
+        elif size == 8:
+            bytes = struct.pack(">Q", value & 0xffffffffffffffff)
+        self.writeMemory(addr, bytes)
+
+    def readMemSignedValue(self, addr, size):
+        #FIXME: Remove byte check and possibly half-word check.  (possibly all but word?)
+        #FIXME: Handle endianness
+        bytes = self.readMemory(addr, size)
+        if bytes == None:
+            return None
+        if size == 1:
+            return struct.unpack("b", bytes)[0]
+        elif size == 2:
+            return struct.unpack(">h", bytes)[0]
+        elif size == 4:
+            return struct.unpack(">l", bytes)[0]
+
+    def integerSubtraction(self, op):
+        """
+        Do the core of integer subtraction but only *return* the
+        resulting value rather than assigning it.
+        (allows cmp and sub to use the same code)
+        """
+        # Src op gets sign extended to dst
+        #FIXME account for same operand with zero result for PDE
+        ssize = op.opers[0].tsize
+        dsize = op.opers[1].tsize
+        src = self.getOperValue(op, 0)
+        dst = self.getOperValue(op, 1)
+
+        if dst == None or src == None:
+            self.undefFlags()
+            return None
+
+        return self.intSubBase(dst, dsize, src, ssize)
+
+    def intSubBase(self, src1, dsize, src2, ssize):
+        # So we can either do a BUNCH of crazyness with xor and shifting to
+        # get the necessary flags here, *or* we can just do both a signed and
+        # unsigned sub and use the results.
+
+        usrc = e_bits.unsigned(src1, ssize)
+        udst = e_bits.unsigned(src2, dsize)
+
+        ssrc = e_bits.signed(src1, ssize)
+        sdst = e_bits.signed(src2, dsize)
+
+        ures = udst - usrc
+        sres = sdst - ssrc
+
+        self.setFlag(self.EMU_H, e_bits.is_signed_half_carry(ures, dsize, udst))
```

Comment:
Perhaps a set of flags *we* define within the function ( and return as a dict full of answers/flags etc ) which an architecture may simply translate to their flag states?